node/task.ts

```diff
@@ -1775,6 +1803,271 @@ export class CodeCoverageEnabler {
+ * 
+ * @param path      Local directory path.
+ * @returns         void
+ */
+export function prependPath(path: string) {
+    assertAgent("2.115.0");
+    command("task.prependpath", null, path);
+}
+
+/**
+ * Upload and attach summary markdown to current timeline record.
+ * This summary shall be added to the build/release summary and 
+ * not available for download with logs.
+ * 
+ * @param path      Local directory path.
+ * @returns         void
+ */
+export function uploadSummary(path: string) {
+    command("task.uploadsummary", null, path);
+}
+
+/**
+ * Upload and attach attachment to current timeline record.
+ * These files are not available for download with logs.
+ * These can only be referred to by extensions using the type or name values. 
+ * 
+ * @param type      Attachment type.
+ * @param name      Attachment name.
+ * @param path      Attachment path.
+ * @returns         void
+ */
+export function addAttachment(type: string, name: string, path: string) {
+    command("task.addattachment", { "type": type, "name": name }, path);
+}
+
+/**
+ * Set an endpoint field with given value.
+ * Value updated will be retained in the endpoint for 
+ * the subsequent tasks that execute within the same job.
+ * 
+ * @param id      Endpoint id.
+ * @param field   FieldType enum of AuthParameter, DataParameter or Url.
+ * @param key     Key.
+ * @param value   Value for key or url.
+ * @returns       void
+ */
+export function setEndpoint(id: string, field: FieldType, key: string, value: string) {
+    command("task.setendpoint", { "id": id, "field": field, "key": key }, value);
+}
+
+/**
+ * Set progress and current operation for current task.
+ * 
+ * @param percent           Percentage of completion.
+ * @param currentOperation  Current pperation. 
+ * @returns                 void
+ */
+export function setProgress(percent: number, currentOperation: string) {
+    command("task.setprogress", { "value": `${percent}` }, currentOperation);
+}
+
+/**
+ * Indicates whether to write the logging command directly to the host or to the output pipeline.
+ *
+ * @param id            Timeline record Guid.
+ * @param parentId      Parent timeline record Guid.
+ * @param recordType    Record type.
+ * @param recordName    Record name.
+ * @param order         Order of timeline record.
+ * @param startTime     Start time.
+ * @param finishTime    End time.
+ * @param progress      Percentage of completion.
+ * @param state         TaskState enum of Unknown, Initialized, InProgress or Completed.
+ * @param result        TaskResult enum of Succeeded, SucceededWithIssues, Failed, Cancelled or Skipped.
+ * @param message       current operation
+ * @returns             void
+ */
+export function logDetail(id: string, parentId?: string, recordType?: string,
+    recordName?: string, order?: number, startTime?: string, finishTime?: string,
+    progress?: number, state?: TaskState, result?: TaskResult, message?: string) {
+    var properties = { "id": id };
```

Comment:
I probably should have been clearer here - we really only need to make this optional in the function declaration, the [check here](https://github.com/Microsoft/azure-pipelines-task-lib/blob/9fdd82cd724c61aa1615e732a20d3a5ffce67397/node/taskcommand.ts#L34) should take care of the rest. So this can still just be
```
export function logDetail(id: string, message: string, parentId?: string, recordType?: string,
    recordName?: string, order?: number, startTime?: string, finishTime?: string,
    progress?: number, state?: TaskState, result?: TaskResult) {
    const properties = {
        "id": id,
        "parentid": parentId,
        "type": recordType,
        "name": recordName,
        "order": order.toString(),
        "starttime": startTime,
        "finishtime": finishTime,
        "progress": progress.toString(),
        "state": TaskState[state],
        "result": TaskResult[result]
    };

    command("task.logdetail", properties, message);
}
```

Sorry, I know I'm being a little picky on style

Reply:
You are right to be picky about it.
I initially tried what you suggest, however, non passed parameters will be set to undefined, thus I was scared on how will a undefined string (in some cases) be processed later by the engine.

Check this:
![image](https://user-images.githubusercontent.com/3026551/49092977-7eedd500-f263-11e8-9f66-e0e7ce43fd03.png)

Reply:
Also, in case of the log detail, is the message also optional?

Reply:
I guess you could have a logDetail without a message, I think it probably makes sense to leave it as required though and the user can pass an explicit null or empty string if they want to omit it. My reasoning here is that unlike ```properties```, [there is typing](https://github.com/Microsoft/azure-pipelines-task-lib/blob/9fdd82cd724c61aa1615e732a20d3a5ffce67397/node/internal.ts#L277) on ```message``` in the ```command``` function. I think this might cause issues if strict type checking is turned on (though I'm not 100% sure).

Reply:
Even if undefined may be handled correctly, in case of strings, a check needs to be preformed first, otherwise a type request ([key: string]: string) will contain a string entry "undefined" which may be assumed later to be the value for a given key.

Reply:
Ah, I see the issue, good catch! Maybe to fix this we can do something like:
```
export function logDetail(id: string, message: string, parentId?: string, recordType?: string,
    recordName?: string, order?: number, startTime?: string, finishTime?: string,
    progress?: number, state?: TaskState, result?: TaskResult) {
    const properties = {
        "id": id,
        "parentid": parentId,
        "type": recordType,
        "name": recordName,
        "order": order ? order.toString() : undefined,
        "starttime": startTime,
        "finishtime": finishTime,
        "progress": progress ? progress.toString() : undefined,
        "state": state ? TaskState[state] : undefined,
        "result": result ? TaskResult[result] : undefined
    };

    command("task.logdetail", properties, message);
}
```

That way any undefined values map correctly to an undefined value in the dictionary and are ignored by ```toString```? I'm mostly just looking for a way to make what is happening a little simpler/less verbose.

Reply:
I like your solution better, I'll fix it right away.